stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -123,25 +123,30 @@ private ClientCalls() {}
   public static <ReqT, RespT> RespT blockingUnaryCall(
       Channel channel, MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, ReqT req) {
     ThreadlessExecutor executor = new ThreadlessExecutor();
+    boolean interrupt = false;
     ClientCall<ReqT, RespT> call = channel.newCall(method, callOptions.withExecutor(executor));
     try {
       ListenableFuture<RespT> responseFuture = futureUnaryCall(call, req);
       while (!responseFuture.isDone()) {
         try {
           executor.waitAndDrain();
         } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw Status.CANCELLED
-              .withDescription("Call was interrupted")
-              .withCause(e)
-              .asRuntimeException();
+          interrupt = true;
+          call.cancel("Thread interrupted", e);
+          // Now wait for onClose() to be called, so interceptors can clean up
         }
       }
       return getUnchecked(responseFuture);
     } catch (RuntimeException e) {
+      // Something very bad happened. All bets are off; it may be dangerous to wait for onClose().
```

Comment:
Wondering how this fix ensures the `onClose()` is run if an exception is thrown on line 129 by `futureUnaryCall()`?  Does this mean it doesn't attempt to run the `onClose()`?

Reply:
Apart from `onClose()`, the executor may also have other listener methods (i.e. `onReady()`, etc.) scheduled.  I'm guessing it's dangerous to run listener methods other than `onClose()` at this point?  Is there a way you can just run the `onClose()`?

Reply:
Because we wait for `responseFuture.isDone()`, then we only call `getUnchecked()` after `onClose()` has been called. The rest of the methods should not throw; if they do, we do a minor amount of clean up, but the system is in a broken state already. We can't run onClose directly because we don't have a reference to the right Listener; only the Channel has that.

`onClose()` is guaranteed to be the last listener method called. So things like `onReady()` shouldn't be a problem.